### PR TITLE
Add separate HTTP and MQTT endpoints for Cumulocity

### PR DESCRIPTION
**Pull request by Bravo555**
## Proposed changes

This PR adds `c8y.http` and `c8y.mqtt` config options intended to override `c8y.url` option in HTTP and MQTT contexts respectively. This is necessary for when HTTP and MQTT servers are present under different domain names. These settings are optional and are derived from `c8y.url` if not present.

In code, settings `C8yHttpSetting` and `C8yMqttSetting` were added. Their `query()` methods were made to look whether `c8y.http/mqtt` option was defined first, and if not, they use the value of `c8y.url`. The old `C8yUrlSetting` accessor was left as it is because changing its behaviour would break setting/listing config options, in the future it should be made deprecated and phased out.

Previously used for deserializing `ConnectUrl` type does not allow specifying ports in host strings. To allow for deserializing with ports optionally specified, as well as to know upfront what default port we use in case it is not specified, and still handle possible parsing errors, `HostPort` type was created, which uses a const type parameter to specify which default port we expect:

[from doctest]:
```rust
// use a fallback port if not present in string
let http = HostPort::<443>::try_from("my-tenant.cumulocity.com".to_string()).unwrap();
assert_eq!(http.port(), Port(443));

// allow port to be overridden using standard `:PORT` notation
let http = HostPort::<443>::try_from("my-tenant.cumulocity.com:8080".to_string()).unwrap();
assert_eq!(http.port(), Port(8080));

// return error for malformed host strings
assert!(HostPort::<443>::try_from("my-tenant.cumulocity.com:8080:443".to_string()).is_err());
assert!(HostPort::<443>::try_from(":8080".to_string()).is_err());
assert!(HostPort::<443>::try_from("[:::1]:8080".to_string()).is_err());
```

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactoring that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
- https://github.com/thin-edge/thin-edge.io/issues/1931

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [x] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

I'd like to hear some feedback regarding:

- **naming**: while better than e.g. `c8y.url`/`C8yUrlSetting`, which are not really URLs, because they can't contain scheme, port, query, fragments, etc. so the name is I think a bit misleading, `HostPort`, while accurate, is maybe a bit awkward. I thought of naming it `ServiceHost`, because really we use it to connect to a known service (HTTP or MQTT) on a certain host and port.
- **structure** while using const generics we can easily create `HostPort` with any expected port, just specifying the port inline kind of smells like magic numbers. We could do an alternative with a trait + marker structs with associated const:

  ```rust
  trait Service {
      const FALLBACK_PORT: u16;
  }
  
  struct Http;
  impl Service for Http {
      const FALLBACK_PORT: u16 = 443;
  }
  
  struct Mqtt;
  impl Service for Mqtt {
      const FALLBACK_PORT: u16 = 8883;
  }
  
  struct HostPort2<S: Service> {
      input: String,
      host: url::Host,
      port: Port,
      _phantomdata: PhantomData<S>,
  }
  
  impl<S: Service> HostPort2<S> {
      /// Returns the port number.
      ///
      /// If `:PORT` suffix was present when deserializing, this specified port is
      /// used. If not, a fallback port from the const type parameter is used.
      pub fn port(&self) -> Port {
          self.port
      }
  }
  
  impl<S: Service> TryFrom<String> for HostPort2<S> {
      type Error = ParseHostnamePortError;
  
      fn try_from(s: String) -> Result<Self, Self::Error> {
          let (host, port) = if let Some((hostname, port)) = s.split_once(':') {
              let port = Port(port.parse()?);
              let hostname: Host<String> = Host::parse(hostname)?;
              (hostname, port)
          } else {
              let hostname: Host<String> = Host::parse(&s)?;
              (hostname, Port(S::FALLBACK_PORT))
          };
  
          Ok(HostPort2 {
              input: s,
              host,
              port,
              _phantomdata: PhantomData,
          })
      }
  }
  ```

  And then we use it the following way:
  ```rust
   let http: HostPort2<Http> =
              HostPort2::try_from("my-tenant.cumulocity.com".to_string()).unwrap();
  ```

- **tests** are they sufficient?

### CODE REVIEW

File: crates/common/tedge_config/src/tedge_config_cli/models/host_port.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+///
+/// // return error for malformed host strings
+/// assert!(HostPort::<443>::try_from("my-tenant.cumulocity.com:8080:443".to_string()).is_err());
+/// assert!(HostPort::<443>::try_from(":8080".to_string()).is_err());
+/// assert!(HostPort::<443>::try_from("[:::1]:8080".to_string()).is_err());
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(into = "String", try_from = "String")]
+pub struct HostPort<const P: u16> {
+    input: String,
```
[the remark this answers is not in the document]
Comment (Bravo555) [reply crates/common/tedge_config/src/tedge_config_cli/models/host_port.rs:37]: We don't want to rewrite the value the user provided when we serialize: e.g. if in the config we have `c8y.http = tenant.cumulocity.com` we don't want to add a port, and if we have `c8y.http = tenant.cumulocity.com:443`, then we don't want to remove the port. That way we can save the original string that we received during deserialization and just return it when we serialize. The alternative would be to save a bool like `was_port_provided` and in `as_str()` return different strings based on whether there was a port specified, but by keeping the original string we also avoid allocations caused by `format!` on a method which is still used by some application code. So IMO it makes logic simpler and code faster although maybe I'm just micro-optimizing.

**Comment by Bravo555** [comment]
I reminded myself that there exists a `#[deprecated]` attribute. I marked `C8yUrlSetting` as deprecated, which makes it so that if anyone tries to use it, a warning appears with a message to use `C8yHttpSetting` or `C8yMqttSetting` instead.

File: crates/common/tedge_config/src/tedge_config_cli/models/host_port.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+/// assert_eq!(http.port(), Port(8080));
+///
+/// // return error for malformed host strings
+/// assert!(HostPort::<443>::try_from("my-tenant.cumulocity.com:8080:443".to_string()).is_err());
+/// assert!(HostPort::<443>::try_from(":8080".to_string()).is_err());
+/// assert!(HostPort::<443>::try_from("[:::1]:8080".to_string()).is_err());
+/// ```
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(into = "String", try_from = "String")]
+pub struct HostPort<const P: u16> {
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/models/host_port.rs:36]: I'm okay to define the default port using a `const` generic.

- First, because using a `Service` trait and phantom values to attach a default port to a struct seems a bit too involved for little gains. I'm not sure how this will work will the derive traits (Debug, Clone, PartialEq, Eq, Serialize, Deserialize).
- Second, because we can improve clarity by adding constants as `const MQTTS : u16 = 8883` and get nicer types as `mqtt_host: HostPort<MQTT>`. 

About the type name.
- I agree that `HostPort` sounds a bit heavy, but at least this conveys something.
- You propose `ServiceHost` as an alternative, I would prefer `ServiceAddress`.
- I let you make a choice.

**Comment by didier-wenzek** [comment]
> I reminded myself that there exists a `#[deprecated]` attribute. I marked `C8yUrlSetting` as deprecated, which makes it so that if anyone tries to use it, a warning appears with a message to use `C8yHttpSetting` or `C8yMqttSetting` instead.

Indeed a good idea. Simple and effective.

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs [hunk 3]
```diff
@@ -174,6 +175,64 @@ impl ConfigSettingAccessor<C8yUrlSetting> for TEdgeConfig {
...
+    #[allow(deprecated)]
+    fn query(&self, _setting: C8yHttpSetting) -> ConfigSettingResult<HostPort<443>> {
+        self.data
+            .c8y
+            .http
+            .as_ref()
+            .cloned()
+            .or(self.data.c8y.url.as_ref().cloned().map(ConnectUrl::into))
+            .ok_or(ConfigSettingError::ConfigNotSet {
+                key: C8yUrlSetting::KEY,
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:188]: Wouldn't be better to tell that `C8yHttpSetting` is not set rather than `C8yUrlSetting`?

```suggestion
                key: C8yHttpSetting::KEY,
```

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs [hunk 4]
```diff
@@ -174,6 +175,64 @@ impl ConfigSettingAccessor<C8yUrlSetting> for TEdgeConfig {
...
+    #[allow(deprecated)]
+    fn query(&self, _setting: C8yMqttSetting) -> ConfigSettingResult<HostPort<8883>> {
+        self.data
+            .c8y
+            .mqtt
+            .as_ref()
+            .cloned()
+            .or(self.data.c8y.url.as_ref().cloned().map(ConnectUrl::into))
+            .ok_or(ConfigSettingError::ConfigNotSet {
+                key: C8yUrlSetting::KEY,
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:217]: Wouldn't be better to tell that `C8yMqttSetting` is not set rather than `C8yUrlSetting`?

```suggestion
                key: C8yMqttSetting::KEY,
```

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config_dto.rs [hunk 5]
```diff
@@ -81,6 +81,14 @@ pub(crate) struct CumulocityConfigDto {
     #[doku(example = "your-tenant.cumulocity.com", as = "String")]
     pub(crate) url: Option<ConnectUrl>,
 
+    /// HTTP Endpoint for the Cumulocity tenant, with optional port.
+    #[doku(example = "http.your-tenant.cumulocity.com:1234", as = "String")]
+    pub(crate) http: Option<HostPort<443>>,
+
+    /// MQTT Endpoint for the Cumulocity tenant, with optional port.
+    #[doku(example = "mqtt.your-tenant.cumulocity.com:1234", as = "String")]
+    pub(crate) mqtt: Option<HostPort<8883>>,
```
Comment (didier-wenzek) [inline crates/common/tedge_config/src/tedge_config_cli/tedge_config_dto.rs:90]: I would define somewhere 2 constants for the default HTTP and MQTT ports.
```suggestion
    /// HTTP Endpoint for the Cumulocity tenant, with optional port.
    #[doku(example = "http.your-tenant.cumulocity.com:1234", as = "String")]
    pub(crate) http: Option<HostPort<HTTP>>,

    /// MQTT Endpoint for the Cumulocity tenant, with optional port.
    #[doku(example = "mqtt.your-tenant.cumulocity.com:1234", as = "String")]
    pub(crate) mqtt: Option<HostPort<MQTT>>,
```

**Review by didier-wenzek — COMMENTED** [review]
Nicely done. I only have some suggestions, but would be happy to merge it like this.

File: crates/common/tedge_config/src/tedge_config_cli/models/host_port.rs (hunk 2 above)
Comment (Bravo555) [reply crates/common/tedge_config/src/tedge_config_cli/models/host_port.rs:36]: > - You propose `ServiceHost` as an alternative, I would prefer `ServiceAddress`.

`ServiceAddress` doesn't sit well with me, because I think DNS names are not addresses in and of themselves, we look them up to eventually resolve them to addresses.

So while maybe not the most graceful name, we can stay with `HostPort`. It can always be changed and the documentation explains its workings to a sufficient degree.

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs (hunk 3 above)
Comment (Bravo555) [reply crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:188]: My rationale is that because `C8yHttpSetting` and `C8yMqttSetting` inherit their values from `c8y.url`, I think of them more as optional overrides rather than settings in their own right. When both `c8y.http` and `c8y.mqtt` are set, then indeed `c8y.url` is not necessary. However, if this error is received by the user, the suggestion to set `c8y.http` is helpful only if `c8y.mqtt` is also already set, which if it isn't, the user will receive a second error.

So in short, IMHO `c8y.url` is sort of "required", and in the error message it makes more sense to tell the user to set a required setting, instead of an optional one.

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs (hunk 4 above)
Comment (Bravo555) [reply crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:217]: responded in the comment for `C8yHttpSetting`

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config_dto.rs (hunk 5 above)
Comment (Bravo555) [reply crates/common/tedge_config/src/tedge_config_cli/tedge_config_dto.rs:90]: I agree, simple and sufficient to reduce the "magic number" code smell

File: crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs (hunk 3 above)
Comment (didier-wenzek) [reply crates/common/tedge_config/src/tedge_config_cli/tedge_config.rs:188]: It makes sense. Ideally, it would be better to tell the user to set either one or the other. But this is not necessary for now.

**Review by didier-wenzek — APPROVED** [review]
Approved
